src/test/app/AMM_test.cpp

```diff
@@ -0,0 +1,6933 @@
+                    XRP(12000), USD(12000), IOUAmount{11999880}));
+                // Discounted trade
+                for (int i = 0; i < 10; ++i)
+                {
+                    ammAlice.deposit(carol, USD(100));
+                    ammAlice.withdraw(carol, USD(100));
+                    ammAlice.deposit(bob, USD(100));
+                    ammAlice.withdraw(bob, USD(100));
+                }
+                BEAST_EXPECT(ammAlice.expectBalances(
+                    XRP(12000), USD(12000), IOUAmount{11999880}));
+                auto const danTokens = ammAlice.getLPTokensBalance(dan);
+                auto const ammTokens = ammAlice.getLPTokensBalance();
+                // Trade with the fee
+                for (int i = 0; i < 10; ++i)
+                {
+                    ammAlice.deposit(dan, USD(100));
+                    ammAlice.withdraw(dan, USD(100));
+                }
+                auto const danFees =
+                    danTokens - ammAlice.getLPTokensBalance(dan);
+                BEAST_EXPECT(ammAlice.expectBalances(
+                    XRP(12000), USD(12000), ammTokens - danFees));
+                // Discounted payment
+                ammAlice.deposit(carol, USD(100));
+                auto tokens = ammAlice.getLPTokensBalance();
+                BEAST_EXPECT(
+                    ammAlice.expectBalances(XRP(12000), USD(12100), tokens));
+                env(pay(carol, bob, USD(100)), path(~USD), sendmax(XRP(110)));
+                env.close();
+                BEAST_EXPECT(
+                    ammAlice.expectBalances(XRP(12100), USD(12000), tokens));
+                // Payment with the fee
+                env(pay(alice, carol, XRP(100)), path(~XRP), sendmax(USD(110)));
+                env.close();
+                BEAST_EXPECT(ammAlice.expectBalances(
+                    XRP(12000),
+                    STAmount{USD, UINT64_C(121010101010101), -10},
+                    tokens));
+                // Auction slot expired, no discounted fee
+                ammAlice.withdraw(
+                    carol, STAmount{USD, UINT64_C(10101010101), -10});
+                tokens = ammAlice.getLPTokensBalance();
+                BEAST_EXPECT(
+                    ammAlice.expectBalances(XRP(12000), USD(12100), tokens));
+                env.close(seconds(TOTAL_TIME_SLOT_SECS + 1));
+                // clock is parent's based
+                env.close();
+                env(pay(carol, bob, USD(100)), path(~USD), sendmax(XRP(110)));
+                env.close();
+                BEAST_EXPECT(ammAlice.expectBalances(
+                    XRPAmount(12101010102), USD(12000), tokens));
+            },
+            std::nullopt,
+            1000);
+
+        // Bid tiny amount
+        testAMM([&](AMM& ammAlice, Env&) {
+            // Can bid a tiny amount
+            auto const tiny = Number{STAmount::cMinValue, STAmount::cMinOffset};
+            ammAlice.bid(alice, IOUAmount{tiny});
```

Comment:
Makes sense. I just wonder if it would be worth having that situation return a failure code (something like `terALREADY_HOLD_SLOT`) when the ledger is open to prevent burning the fee. But then the user would have to reuse the same sequence. I don't know enough about arbitrage workflows to know if it would be worth it.